### PR TITLE
Rewrite of the `Rhythm` Skill within osu!taiko

### PR DESCRIPTION
**Pull request by Lawtrohux**
# Rewrite of Rhythm within osu!taiko

## Ratio Difficulty
In this rework, rhythm difficulty is based on the fact that more complex interval ratios are rhythmically more difficult. In general, we can define more complex ratios as fractions that have larger denominators in their simplified form. For example, 3/5 is more difficult than 1/2. With these assumptions, we construct a sine wave-based function that penalises the specified denominator.

Credits to YaniFR for porting these changes over to the new pp-dev branch.

---

## Rhythm Changes
- Introduction of three new rhythm data classes:
  - `EvenHitObjects`
  - `EvenPatterns`
  - `EvenRhythm`
- Intervals are accessed via the `IHasInterval` interface.

### Move Rhythm into an Evaluator for per-object difficulty.
- Penalises consistent intervals within rhythm.
- Gives bonuses to near-1 ratio's, and penalises very-near-1 ratio's.
- Penalises exceedingly long gaps between awkward rhythm (primarily found in low SR difficulties).

---

## Where can I view changes? 
For a dev-based writeup on the Rhythm and Reading Skills, these are viewable here:  
[Dev Writeup](https://docs.google.com/document/d/1-k96U-3-77CSyWNbXxa3cf2AXeOIxVh8yoXowlIYc5g/edit?usp=sharing)

For a very general writeup meant for end users, view these here:  
[General Writeup](https://www.ltca.pro/osu!taiko/ppv3)

For testing of values, this can be viewed on huismetbenen:  
[huismetbenen](https://pp.huismetbenen.nl/rankings/players/taiko-ppv3)

## Some already present issues
Due to strain chunking, rate changing some maps can cause values to go up or down (within 0.01* margin) depending on very small rate changes, this is also present within the `MarginOfError` with rhythm, as slight interval changes (outside of 5ms) move pattern grouping. This doesn't contribute to any openings for abuse. This is something outside of the constraint of what this pr aims to achieve, and as rate changes stay unranked, and this depends on changing how strains chunk.

### CODE REVIEW

**Review by YaniFR — APPROVED** [review]
will need performance point balancing but we're leaving that for another pr, i reviewed the code with ltca in vc for about an hour before this approval.

File: osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Rhythm/Data/EvenHitObjects.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+using System.Collections.Generic;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Rhythm.Data
+{
+    /// <summary>
+    /// Represents a group of <see cref="TaikoDifficultyHitObject"/>s with no rhythm variation.
+    /// </summary>
+    public class EvenHitObjects : EvenRhythm<TaikoDifficultyHitObject>, IHasInterval
```
Comment (stanriders) [inline osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Rhythm/Data/EvenHitObjects.cs:12]: Wouldn't it make more sense to call it something like `SimilarRhythmHitObjects`? `Even` sounds like its related to a pattern size (even/odd)

File: osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Rhythm/Data/EvenPatterns.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Rhythm.Data
+{
+    /// <summary>
+    /// Represents <see cref="EvenHitObjects"/> grouped by their <see cref="EvenHitObjects.StartTime"/>'s interval.
+    /// </summary>
+    public class EvenPatterns : EvenRhythm<EvenHitObjects>
```
Comment (stanriders) [inline osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Rhythm/Data/EvenPatterns.cs:12]: Same as above

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs [hunk 3]
```diff
@@ -129,14 +139,14 @@ protected override DifficultyAttributes CreateDifficultyAttributes(IBeatmap beat
             {
                 StarRating = starRating,
                 Mods = mods,
-                RhythmDifficulty = rhythmRating,
+                RhythmDifficulty = rhythmRating * 8,
```
Comment (stanriders) [inline osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs:142]: That doesn't look intentional. If it is it needs an explanation

**Review by stanriders — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs (hunk 3 above)
Comment (Lawtrohux) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs]: oops... that was for a last minute test post-review

**Review by stanriders — DISMISSED** [review]

**Review by stanriders — APPROVED** [review]

**Pull request merged**